staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go

```diff
@@ -0,0 +1,212 @@
+	test "k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/testing"
+	"k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/testing/clock"
+)
+
+type uniformScenario []uniformClient
+
+type uniformClient struct {
+	hash          uint64
+	nThreads      int
+	nCalls        int
+	execDuration  time.Duration
+	thinkDuration time.Duration
+}
+
+// exerciseQueueSetUniformScenario.  Simple logic, only works if each
+// client's offered load is at least as large as its fair share of
+// capacity.
+func exerciseQueueSetUniformScenario(t *testing.T, qs fq.QueueSet, sc uniformScenario,
+	totalDuration time.Duration, expectPass bool, expectedAllRequests bool,
+	clk *clock.FakeEventClock, counter counter.GoRoutineCounter) {
+
+	now := time.Now()
+	t.Logf("%s: Start", clk.Now().Format("2006-01-02 15:04:05.000000000"))
+	integrators := make([]test.Integrator, len(sc))
+	var failedCount uint64
+	for i, uc := range sc {
+		integrators[i] = test.NewIntegrator(clk)
+		for j := 0; j < uc.nThreads; j++ {
+			counter.Add(1)
+			go func(i, j int, uc uniformClient, igr test.Integrator) {
+				for k := 0; k < uc.nCalls; k++ {
+					ClockWait(clk, counter, uc.thinkDuration)
+					for {
+						tryAnother, execute, afterExecute := qs.Wait(context.Background(), uc.hash)
+						t.Logf("%s: %d, %d, %d got q=%v, e=%v", clk.Now().Format("2006-01-02 15:04:05.000000000"), i, j, k, tryAnother, execute)
+						if tryAnother {
+							continue
+						}
+						if !execute {
+							atomic.AddUint64(&failedCount, 1)
+							break
+						}
+						igr.Add(1)
+						ClockWait(clk, counter, uc.execDuration)
+						afterExecute()
+						igr.Add(-1)
+						break
+					}
+				}
+				counter.Add(-1)
+			}(i, j, uc, integrators[i])
+		}
+	}
+	lim := now.Add(totalDuration)
+	clk.Run(&lim)
+	clk.SetTime(lim)
+	t.Logf("%s: End", clk.Now().Format("2006-01-02 15:04:05.000000000"))
+	results := make([]test.IntegratorResults, len(sc))
+	var sumOfAvg float64
+	for i := range sc {
+		results[i] = integrators[i].GetResults()
+		sumOfAvg += results[i].Average
+	}
+	idealAverage := sumOfAvg / float64(len(sc))
+	passes := make([]bool, len(sc))
+	allPass := true
+	for i := range sc {
+		relDiff := (results[i].Average - idealAverage) / idealAverage
+		passes[i] = math.Abs(relDiff) <= 0.1
+		allPass = allPass && passes[i]
+	}
+	for i := range sc {
+		if allPass != expectPass {
+			t.Errorf("Class %d got an Average of %v but the ideal was %v", i, results[i].Average, idealAverage)
+		} else {
+			t.Logf("Class %d got an Average of %v and the ideal was %v", i, results[i].Average, idealAverage)
+		}
+	}
+
+	clk.Run(nil)
+	if expectedAllRequests && failedCount > 0 {
+		t.Errorf("Expected all requests to be successful but got %v failed requests", failedCount)
+	} else if !expectedAllRequests && failedCount == 0 {
+		t.Errorf("Expected failed requests but all requests succeeded")
+	}
+}
+
+// TestNoRestraint should fail because the dummy QueueSet exercises no control
```

Comment:
What exactly should fail? Not the whole test, or we wouldn't check it in?

Reply:
This was the first test function written, before we had a real implementation of QueueSet.  Now that we do, this test is relatively uninteresting.  To be fair, it _does_ test the test.